### ISSUE TITLE
Bug: When screenshotting a very long element, text at the bottom of the image is saved with a black background

### ISSUE THREAD

**Issue by void4** [mozilla/geckodriver#2133]
## System

* Version: [geckodriver 0.33.0 (a80e5fd61076 2023-04-02 18:31 +0000)](https://github.com/mozilla/geckodriver/releases/tag/v0.33.0)
* Platform: Ubuntu 20.04.6 LTS
* Firefox: Mozilla Firefox 117.0 is installed on the system
* Selenium: 4.1.3 (pretty old (?), but newer versions don't work in my specific setup)

## Problem

When taking a screenshot of a very long (vertically large) element on a page (e.g. https://icebergcharts.com/i/The_Complete_Geometry_Dash), text at the bottom of the page gets a black background in the image that isn't there on the original page.

![image](https://github.com/mozilla/geckodriver/assets/11392207/901b5732-415c-4281-a426-1cf035f440a2)

## Testcase

```python
from selenium import webdriver
from db import cwd
import os
from time import sleep

DEBUG = False

def snap(url, target):

    geckopath = "/home/test/geckodriver"

    opt = webdriver.FirefoxOptions()
    opt.add_argument('-headless')
    opt.add_argument("--width=1920")
    opt.add_argument("--height=1080")

    if DEBUG:
        print("starting webdriver")

    driver = webdriver.Firefox(executable_path=geckopath, options=opt)
    try:
        print("getting url")
        driver.get(url)

        print("sleeping")
        sleep(3)

        print("taking screenshot")
        element = driver.find_element_by_id("iceberg")
        element_png = element.screenshot_as_png
        with open(target, "wb") as file:
            file.write(element_png)

    finally:
        driver.quit()

snap("https://icebergcharts.com/i/The_Complete_Geometry_Dash", "snapshot.png")

```

**Comment by whimboo** [comment]
For me only the very last line, which is the footer, is not getting a background. Is that what you mean? I do not see 4 rows. Does the screenshot work when you only screenshot this particular element? If you could test that I would appreciate. Thanks!

**Comment by void4** [comment]
Ah, sorry, should have been more specific, I mean this:
![image](https://github.com/mozilla/geckodriver/assets/11392207/877678bd-6957-485d-a9d2-564ed866cd0d)

**Comment by whimboo** [comment]
I can see that now as well. The internal print as PDF feature of Firefox doesn't seem to produce that. Interesting is that we basically use the exact same API for it. Would you mind to maybe reduce this very huge test case? It would be much easier to investigate or point to a specific known bug when the HTML / CSS would be minimized.

**Comment by void4** [comment]
I can, but give me a few days. There seem to be more text-related problems when it comes to long pages, such as incompletely rendered text (see the "Exen is legi?" entry on the left), and also text not appearing at all on *very* long pages.

**Comment by whimboo** [comment]
Sure no worries. It most likely wont be a priority right now for geckodriver, but in case its a bug in Gecko itself we could more easily find a solution then. Thanks!

**Comment by whimboo** [comment]
@void4 I wanted to check back if you had a chance to find / create a minimized testcase for this scenario. If not we may have to work with the full page to figure out what's wrong. Thanks.

**Comment by void4** [comment]
No, not yet - I'm currently focusing on other things, so this isn't a priority for me, though it would be nice to see a solution

**Comment by void4** [comment]
Okay, since the issue continues to occur, I investigated it a bit further and produced a small example! This bug has something to do with CSS text-shadow rendering on long pages. Here is how you can reproduce it:

Save the below as a Python script, change the geckopath variable value to point to your geckodriver path and run it. It creates an HTML string with the bug and then directly opens it with selenium to take a screenshot, saving it as screenshot.png

```python
from selenium import webdriver
from selenium.webdriver.firefox.service import Service

# TODO change to your geckodriver path here
geckopath = "/home/one/blacktext/geckodriver"

# HTML creation part
# if you change the text-shadow color, the color of the rectangles behind the text changes too
# the bug also happens when changing the div to a pre
html = '<div id="test" style="text-shadow: 1px 1px 1px red;">'
html += "AAAAAAAAAAAAAAA<br>" * 600
html += "</div>"

# Screenshot part
opt = webdriver.FirefoxOptions()

# Changing/omitting these does not seem to make a difference
opt.add_argument("--width=1920")
opt.add_argument("--height=10080")

service = Service(executable_path=geckopath)

driver = webdriver.Firefox(service=service, options=opt)

driver.get("data:text/html;charset=utf-8,"+html)

element = driver.find_element("id", "test")
element_png = element.screenshot_as_png
with open("screenshot.png", "wb") as file:
    file.write(element_png)
```

Example screenshot:

<img width="1904" height="11400" alt="Image" src="https://github.com/user-attachments/assets/369debbf-2fed-4680-a9ac-dd2c772a92df" />

**Comment by void4** [comment]
@whimboo Sorry for taking 2 years ^^
